core/trino-spi/src/main/java/io/trino/spi/block/AbstractRowBlock.java

```diff
@@ -65,6 +65,23 @@ protected AbstractRowBlock(int numFields)
         this.numFields = numFields;
     }
 
+    /**
+     * Processes field {@link Block}s at the given positions using supplied {@link FieldBlocksProcessor}.
+     */
+    public void processFieldBlocks(int[] nonNullPositions, int offset, int length, FieldBlocksProcessor processor)
+    {
+        int[] fieldPositions = new int[length];
+        for (int i = offset; i < offset + length; i++) {
+            int position = nonNullPositions[i];
+            if (isNull(position)) {
+                throw new IllegalArgumentException("Position is null " + position);
+            }
+            fieldPositions[i] = getFieldBlockOffset(position);
+        }
+
+        processor.process(getRawFieldBlocks(), fieldPositions);
```

Comment:
Looks like access to protected method `getRawFieldBlocks` is essentially what you need.
So how about we add a method `copyRowBlockPositions(AbstractRowBlock block, int[] srcPositions, FieldBlocksProcessor[] fieldProcessors, boolean[] rowIsNull, int[] fieldBlockOffsets, int destOffset)` to BlockUtil ? It'll avoid the need to make one pass for getting non-null positions and another pass to get field block offsets of non-null positions. It'll also avoid repeat of `isNull` check on already known to be non-null positions.

Reply:
Yes, what I need is access to `getRawFieldBlocks` (and `fieldBlockOffsets`). Adding methods to `BlockUtil` won't help as `BlockUtil` is `package private` but I could add `copyRowBlockPositions` to the `AbstractRowBlock`.
It has some small performance advantages over the original change but the issue I have with this is that:
1. it  moves `RowPositionsAppender` logic to the `AbstractRowBlock` (e.g. updating offsets). The current `processFieldBlocks` implementation is very simple. Just map the positions and delegate to the processor.
2. it exposes most of the `RowPositionsAppender` private state. This makes it harder to understand the `RowPositionsAppender` code and bloats the `copyRowBlockPositions`  API with multiple parameters.
3. It's hard to see how the `rowIsNull` and `fieldBlockOffsets` parameters are related to the `fieldProcessors`

Reply:
> I don't think Block has any writeXXX methods, has it?

Right it doesn't. Let's just follow the pattern of ColumnarRow and RowBlockEncoding then by defining something in public in io.trino.spi.block package which can internally access the protected methods of RowBlock.

Reply:
> How about making io.trino.spi.block.AbstractRowBlock#getFieldBlockOffset public?

We could, but it would be easier to land the change if we take similar approach as ColumnarRow and RowBlockEncoding and avoid doing that.

Reply:
> similar approach as ColumnarRow and RowBlockEncoding 

Do you have in mind putting something like `RowPositionsAppender` in the `io.trino.spi.block` or some other idea?

Reply:
I'm thinking we add a class `BatchedRowBlockBuilder` in `io.trino.spi.block`. Add API `public void copyPositions(int[] positions, int offset, int length, BatchedRowBlockBuilder rowBlockBuilder)` to `RowBlock`.
RowPositionsAppender still stays in `io.trino.operator.output` and makes use of `BatchedRowBlockBuilder` and new API in `RowBlock`.